Upstream-Status: Pending [Not submitted to upstream yet]
Signed-off-by: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>

From ff2b492bf68337d6cdd296ed61f91304b2da3143 Mon Sep 17 00:00:00 2001
From: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>
Date: Tue, 7 Dec 2021 00:26:37 +0000
Subject: [PATCH] Fix GetVariable() to handle 0 data size

This patch fixes the GetVariable() issue which causes
mm_communicate failure when called with 0 data size. The comm buffer
is set to maximum size when 0 data size request is made to handle the
MM response from the secure world. This is a generic fix but used by
corstone1000.

Signed-off-by: Gowtham Suresh Kumar <gowtham.sureshkumar@arm.com>

%% original patch: 0035-corstone1000-Add-support-in-GetVariable-to-handle-re.patch
---
 lib/efi_loader/efi_variable_tee.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/lib/efi_loader/efi_variable_tee.c b/lib/efi_loader/efi_variable_tee.c
index 9375aa6a63..b7de80cd39 100644
--- a/lib/efi_loader/efi_variable_tee.c
+++ b/lib/efi_loader/efi_variable_tee.c
@@ -637,10 +637,15 @@ efi_status_t efi_get_variable_int(u16 *variable_name, const efi_guid_t *vendor,
 		goto out;
 	}
 
-	/* Trim output buffer size */
+	/* Set output buffer size to maximum value if data_size is 
+	 * larger than the supported value or if efi_get_variable_int()
+	 * is called with data_size set to 0. In the latter case if
+	 * output buffer is not set to maximum, the comm_buf will be 
+	 * smaller than the MM response.
+	 */
 	tmp_dsize = *data_size;
-	if (name_size + tmp_dsize >
-			max_payload_size - MM_VARIABLE_ACCESS_HEADER_SIZE) {
+	if ((name_size + tmp_dsize > max_payload_size - MM_VARIABLE_ACCESS_HEADER_SIZE)
+	    			|| (!tmp_dsize)) {
 		tmp_dsize = max_payload_size -
 				MM_VARIABLE_ACCESS_HEADER_SIZE -
 				name_size;
-- 
2.17.1

